runtime/src/gpu/rocm/gpu-rocm.c

```diff
@@ -27,67 +27,497 @@
+  CUDA_CALL(hipMemcpyDtoH(dst, (hipDeviceptr_t)src, n));
+}
 
-void chpl_gpu_impl_copy_device_to_device(void* dst, const void* src, size_t n) {}
+void chpl_gpu_impl_copy_host_to_device(void* dst, const void* src, size_t n) {
+  assert(chpl_gpu_is_device_ptr(dst));
 
-bool chpl_gpu_impl_is_device_ptr(const void* ptr) {
-  return false;
+  CUDA_CALL(hipMemcpyHtoD((hipDeviceptr_t)dst, src, n));
 }
 
-bool chpl_gpu_impl_is_host_ptr(const void* ptr) {
-  return false;
+void chpl_gpu_impl_copy_device_to_device(void* dst, const void* src, size_t n) {
+  assert(chpl_gpu_is_device_ptr(dst) && chpl_gpu_is_device_ptr(src));
+
+  CUDA_CALL(hipMemcpyDtoD((hipDeviceptr_t)dst, (hipDeviceptr_t)src, n));
+}
+
+
+void* chpl_gpu_impl_comm_async(void *dst, void *src, size_t n) {
+/*  hipStream_t stream;
+  hipStreamCreateWithFlags(&stream, hipStreamNonBlocking);
+  cuMemcpyAsync((hipDeviceptr_t)dst, (hipDeviceptr_t)src, n, stream);
+  return stream;*/
+  assert(false);
+  return NULL;
+}
+
+void chpl_gpu_impl_comm_wait(void *stream) {
+  hipStreamSynchronize((hipStream_t)stream);
+  hipStreamDestroy((hipStream_t)stream);
+}
+
+void* chpl_gpu_mem_array_alloc(size_t size, chpl_mem_descInt_t description,
+                               int32_t lineno, int32_t filename) {
+
+  chpl_gpu_ensure_context();
+
+  CHPL_GPU_DEBUG("chpl_gpu_mem_array_alloc called. Size:%d file:%s line:%d\n", size,
+               chpl_lookupFilename(filename), lineno);
+
+  hipDeviceptr_t ptr = 0;
+  if (size > 0) {
+    chpl_memhook_malloc_pre(1, size, description, lineno, filename);
+#ifdef CHPL_GPU_MEM_STRATEGY_ARRAY_ON_DEVICE
+    if (chpl_gpu_device_alloc) {
+      CUDA_CALL(hipMalloc(&ptr, size));
+    }
+    else {
+//      void* mem = chpl_mem_alloc(size, description, lineno, filename);
+//      CHPL_GPU_DEBUG("\tregistering %p\n", mem);
+//      CUDA_CALL(hipHostRegister(mem, size, hipHostRegisterPortable));
+//      CUDA_CALL(hipHostGetDevicePointer(&ptr, mem, 0));
+      return chpl_malloc(size);
+    }
+#else
+    CUDA_CALL(hipMallocManaged(&ptr, size, hipMemAttachGlobal));
+#endif
+
+    chpl_memhook_malloc_post((void*)ptr, 1, size, description, lineno, filename);
+
+    CHPL_GPU_DEBUG("chpl_gpu_mem_array_alloc returning %p\n", (void*)ptr);
+  }
+  else {
+    CHPL_GPU_DEBUG("chpl_gpu_mem_array_alloc returning NULL (size was 0)\n");
+  }
+
+  return (void*)ptr;
+}
+
+
+void* chpl_gpu_impl_mem_alloc(size_t size) {
+  chpl_gpu_ensure_context();
+
+#ifdef CHPL_GPU_MEM_STRATEGY_ARRAY_ON_DEVICE
+  void* ptr = 0;
+  CUDA_CALL(hipHostAlloc(&ptr, size, 0));
+#else
+  hipDeviceptr_t ptr = 0;
+  CUDA_CALL(hipMallocManaged(&ptr, size, hipMemAttachGlobal));
+#endif
+  assert(ptr!=0);
+
+  return (void*)ptr;
+}
+
+void chpl_gpu_impl_mem_free(void* memAlloc) {
+  chpl_gpu_ensure_context();
+
+  if (memAlloc != NULL) {
+    assert(chpl_gpu_is_device_ptr(memAlloc));
+#ifdef CHPL_GPU_MEM_STRATEGY_ARRAY_ON_DEVICE
+    if (chpl_gpu_impl_is_host_ptr(memAlloc)) {
```

Comment:
Sorry, this asserts that `memAlloc` is a _device_ pointer, after which it checks for whether it's a _host_ pointer? Seems mutually exclusive to me, but maybe I'm missing something

Reply:
Yeah, this looks odd to me too. This code is copied over from the CUDA runtime:

https://github.com/chapel-lang/chapel/blob/e94f7dfe1e7efe33fd7f7cf4cf1ec620b959c2ae/runtime/src/gpu/cuda/gpu-cuda.c#L425

So we should update it there as well. Again, I'm going to spawn off a new issue for this: 

https://github.com/chapel-lang/chapel/issues/21690